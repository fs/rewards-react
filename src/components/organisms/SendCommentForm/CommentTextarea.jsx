import React from 'react';
import ReactTextareaAutocomplete from '@webscopeio/react-textarea-autocomplete';
import '../../../styles/autocomplete.css';
import BonusPossibilitiesService, { POINTS, TAGS } from '../../../services/BonusPossibilitiesService';

export const pointItem = ({ entity: { id, value } }) => (
  <div data-testid="test-point-item" key={id}>{`+â™¥${value}`}</div>
);

export const tagItem = ({ entity: { id, label } }) => <div data-testid="test-tag-item" key={id}>{`#${label}`}</div>;

const CommentTextarea = props => {
  const { onChange, textareaValue } = props;

  const points = BonusPossibilitiesService.getPossibilities(POINTS);
  const tags = BonusPossibilitiesService.getPossibilities(TAGS);

  return (
    <div data-testid="test-textarea-wrapper">
      <ReactTextareaAutocomplete
        className="autocomplete-textarea autocomplete-textarea--comment"
        onChange={onChange}
        value={textareaValue}
        placeholder="Add a comment"
        data-testid="test-textarea"
        loadingComponent={() => <span>Loading</span>}
        name="bonus-text"
        trigger={{
          '+': {
            dataProvider: token => points.filter(point => point.id.includes(token)),
            component: pointItem,
            output: (item, trigger) => `${trigger}${item.value.toString()}`,
          },
          '#': {
            dataProvider: token => tags.filter(tag => tag.label.includes(token)),
            component: tagItem,
            output: (item, trigger) => `${trigger}${item.label}`,
          },
        }}
        minChar="0"
      />
    </div>
  );
};

export default CommentTextarea;
