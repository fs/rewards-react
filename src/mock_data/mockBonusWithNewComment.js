export default {
  comments: [
    {
      id: '610',
      sender: {
        name: 'Nadezhda Kharchuk',
        'allowance-balance': 141,
      },
      text: [
        {
          text: '+1',
          type: 'text',
        },
        {
          text: '#create-awesomeness',
          type: 'tags',
        },
      ],
    },
  ],
  'created-at': '2019-08-28T07:41:29.988Z',
  id: '555',
  points: 1,
  receivers: [
    {
      'allowance-balance': 490,
      'bonus-balance': 97,
      email: 'albert.fazullin@flatstack.com',
      'full-name': 'Albert Fazullin',
      'profile-image-avatar-url':
        'http://doypx5fd26upp.cloudfront.net/assets/default-user-profile_image-b50dbea6a3e28a1ae62538d4b0115b18558fac3ff7dbbee7da800b5216f97070.png',
      username: 'albert.fazullin',
    },
  ],
  sender: {
    name: 'Nadezhda Kharchuk',
    'allowance-balance': 141,
  },
  text: [
    {
      text: '+1',
      type: 'points',
    },
    {
      text: '@albert.fazullin',
      type: 'users',
    },
    {
      text: '#be-curious-never-stop-learning',
      type: 'tags',
    },
  ],
  'total-points': 2,
};
