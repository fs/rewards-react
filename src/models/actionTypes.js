export const UPDATE_USER_LOADING = 'UPDATE_USER_LOADING';
export const UPDATE_USER_ERROR = 'UPDATE_USER_ERROR';
export const UPDATE_USER_SUCCESS = 'UPDATE_USER_SUCCESS';
export const UPDATE_BONUS_LIST_SUCCESS = 'UPDATE_BONUS_LIST_SUCCESS';
export const UPDATE_BONUS_LIST_ERROR = 'UPDATE_BONUS_LIST_ERROR';
export const UPDATE_BONUS_LIST_LOADING = 'UPDATE_BONUS_LIST_LOADING';
