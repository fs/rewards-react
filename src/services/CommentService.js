import api from './ApiService';
import AuthService from './AuthService';

export default class CommentService {
  static async createComment(text, bonusId) {
    const apiUrl = `/user/bonuses/${bonusId}/comments`;
    const token = AuthService.getToken();
    const config = {
      headers: { Authorization: `Bearer ${token}` },
    };

    const bodyParameters = {
      data: {
        type: 'comment-texts',
        attributes: {
          text,
        },
      },
    };

    const response = await api.post(apiUrl, bodyParameters, config);
    return response;
  }
}
