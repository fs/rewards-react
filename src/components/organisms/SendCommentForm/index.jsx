import React, { useState, useContext } from 'react';
import styled from 'styled-components';
import Context from '../../context/Context';
import CommentTextarea from './CommentTextarea';
import Button from '../../atoms/Button';
import HelperIcon from '../../atoms/HelperIcon';
import { UPDATE_BONUS_LIST_AFTER_ADD_COMMENT_SUCCESS } from '../../../models/actionTypes';

const Form = styled.form`
  width: 100%;
  background-color: #f7f7f7;
  border-radius: 5px;
  margin: 0 0 20px;
  padding: 12px 20px;
  box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.1);
  outline: none;
`;

const ErrorContainer = styled.small`
  display: block;
  padding: 5px 10px;
  background-color: #ffe0e0;
  color: #df1616;
  font-weight: bold;
  margin-top: 5px;
`;

const HelperIconsContainer = styled.div`
  display: flex;
`;

const Controls = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  overflow: hidden;
  height: 0;
  transition: height 500ms;
  ${({ isControlsShowing }) =>
    isControlsShowing &&
    `
    height: 55px;
  `};
`;

const regexObj = { points: /\+[1-9]\d*/, userNames: /@\w+/, hashTags: /#\w+/ };

const SendCommentForm = props => {
  const [commentTextareaValue, setCommentTextareaValue] = useState('');
  const [hasError, setHasError] = useState(false);
  const [errorMessage, setErrorMessage] = useState('');
  const [messagePointsIsValid, setMessagePointsIsValid] = useState(false);
  const [messageHashTagIsValid, setMessageHashTagIsValid] = useState(false);
  const [commentButtonText, setCommentButtonText] = useState('Add');
  const [isControlsShowing, setIsControlsShowing] = useState(false);
  const { dispatch } = useContext(Context);

  const validate = (values, regex) => {
    const res = values.find(value => value.match(regex));
    return !!res;
  };

  const getTotalPoints = values => {
    const receiversCount = values.filter(value => value.match(regexObj.userNames)).length;
    const points = values.find(value => value.match(regexObj.points)).slice(1);
    return receiversCount && points ? `+ ${receiversCount * points} Add` : 'Add';
  };

  const updateCommentButton = values => {
    if (validate(values, regexObj.points) && validate(values, regexObj.userNames)) {
      setCommentButtonText(getTotalPoints(values));
    } else {
      setCommentButtonText('Add');
    }
  };

  const handleChange = event => {
    const { value } = event.target;
    const values = value.split(' ');
    setHasError(false);
    setErrorMessage('');
    setCommentTextareaValue(value);
    setMessagePointsIsValid(validate(values, regexObj.points));
    setMessageHashTagIsValid(validate(values, regexObj.hashTags));
    updateCommentButton(values);
  };

  const handleSubmit = async event => {
    event.preventDefault();
    const { onSuccess, commentService, bonusId } = props;

    try {
      const response = await commentService.createComment(commentTextareaValue, bonusId);

      setCommentTextareaValue('');
      onSuccess();
      dispatch({ type: UPDATE_BONUS_LIST_AFTER_ADD_COMMENT_SUCCESS, payload: response.data });
    } catch (error) {
      let parsedErrorMessage;
      if (error.response && error.response.request && error.response.request.response) {
        parsedErrorMessage = JSON.parse(error.response.request.response).errors[0].detail;
      } else {
        console.error(error);
        parsedErrorMessage = 'Error occurred';
      }
      setHasError(true);
      setErrorMessage(parsedErrorMessage);
    }
  };

  const handleIconClick = sign => {
    const text = commentTextareaValue && !commentTextareaValue.endsWith(' ') ? ` ${sign}` : sign;
    setCommentTextareaValue(commentTextareaValue + text);
  };

  return (
    <Form
      tabIndex={0}
      onFocus={() => setIsControlsShowing(true)}
      onBlur={() => setIsControlsShowing(false)}
      onSubmit={handleSubmit}
      data-testid="test-bonus-form"
    >
      <CommentTextarea
        onChange={handleChange}
        textareaValue={commentTextareaValue}
        messagePointsIsActive={messagePointsIsValid}
        messageHashTagsIsActive={messageHashTagIsValid}
      />

      <div data-testid="test-error-container">{hasError && <ErrorContainer>{errorMessage}</ErrorContainer>}</div>

      <Controls isControlsShowing={isControlsShowing}>
        <Button text={commentButtonText} />

        <HelperIconsContainer>
          <HelperIcon
            imgPath={require('../../../images/helper-icon-points.svg')}
            alt="Points"
            isActive={messagePointsIsValid}
            onClick={() => handleIconClick('+')}
          />

          <HelperIcon
            imgPath={require('../../../images/helper-icon-hashtag.svg')}
            alt="Hashtag"
            isActive={messageHashTagIsValid}
            onClick={() => handleIconClick('#')}
          />
        </HelperIconsContainer>
      </Controls>
    </Form>
  );
};

export default SendCommentForm;
