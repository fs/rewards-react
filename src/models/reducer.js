import * as types from './actionTypes';
import bonusParser from '../utils/bonusParser';

const reducer = (state, action) => {
  switch (action.type) {
    case types.UPDATE_USER_SUCCESS:
      return {
        ...state,
        user: action.payload,
        isUserLoading: false,
        hasUserError: false,
      };
    case types.UPDATE_USER_LOADING:
      return {
        ...state,
        isUserLoading: true,
        hasUserError: false,
      };
    case types.UPDATE_USER_ERROR:
      return {
        ...state,
        isUserLoading: false,
        hasUserError: true,
      };
    case types.UPDATE_BONUS_LIST_SUCCESS:
      return {
        ...state,
        bonusList: action.payload,
        isBonusListLoading: false,
        hasBonusListError: false,
      };
    case types.UPDATE_BONUS_LIST_ERROR:
      return {
        ...state,
        hasBonusListError: true,
        isBonusListLoading: false,
      };
    case types.UPDATE_BONUS_LIST_LOADING:
      return {
        ...state,
        isBonusListLoading: true,
        hasBonusListError: false,
      };
    case types.UPDATE_BONUS_LIST_AFTER_ADD_COMMENT_SUCCESS: {
      const newBonus = bonusParser(action.payload);
      const { bonusList } = state;

      const updatedBonusList = bonusList.map(item => {
        if (item.id === newBonus[0].id) {
          return newBonus[0];
        }
        return item;
      });

      console.log('action', updatedBonusList);

      return {
        ...state,
        bonusList: updatedBonusList,
      };
    }
    default:
      return state;
  }
};

export default reducer;
