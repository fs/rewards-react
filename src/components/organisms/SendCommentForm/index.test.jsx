import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import React from 'react';
import { render, fireEvent, wait } from '@testing-library/react';

import CommentService from '../../../services/CommentService';
import Context from '../../context/Context';
import SendCommentForm from '.';

import * as types from '../../../models/actionTypes';

import mockBonusWithNewComment from '../../../mock_data/mockBonusWithNewComment';

jest.mock('../../../services/CommentService');

describe('SendCommentForm', () => {
  const dispatch = jest.fn();
  const expectedBonusId = '555';

  beforeEach(() => {
    jest.resetModules();
    jest.clearAllMocks();

    CommentService.createComment.mockClear();
  });

  test('should call CommentService.createComment on submit', async () => {
    // Arrange
    const expectedCommentCount = '+1';
    const expectedTag = '#create-awesomeness';
    const expectedCommentText = `${expectedCommentCount} ${expectedTag}`;

    const expectedNewBonusAction = {
      type: types.UPDATE_BONUS_LIST_AFTER_ADD_COMMENT_SUCCESS,
      payload: mockBonusWithNewComment,
    };

    const expectedAllowanceBalanceAction = {
      type: types.UPDATE_ALLOWANCE_BALANCE,
      payload: mockBonusWithNewComment.sender['allowance-balance'],
    };

    const mockCreateComment = jest.fn(() => Promise.resolve(mockBonusWithNewComment));

    CommentService.createComment.mockImplementation(mockCreateComment);

    const { getByTestId } = render(
      <Context.Provider value={{ dispatch }}>
        <SendCommentForm onSuccess={() => {}} bonusId={expectedBonusId} />,
      </Context.Provider>,
    );
    const textArea = getByTestId('test-textarea');
    fireEvent.change(textArea, { target: { value: expectedCommentText } });
    const form = getByTestId('test-comment-form');

    // Act
    fireEvent.submit(form);

    // Assert
    await wait(() => {
      expect(CommentService.createComment).toHaveBeenCalledWith(expectedCommentText, expectedBonusId);
    });

    expect(dispatch).toHaveBeenNthCalledWith(1, expectedNewBonusAction);
    expect(dispatch).toHaveBeenNthCalledWith(2, expectedAllowanceBalanceAction);
  });

  test('should call CommentService.createComment on submit and show error message when not enough points', async () => {
    // Arrange
    const expectedCommentCount = '+10000000000';
    const expectedTag = '#create-awesomeness';
    const expectedCommentText = `${expectedCommentCount} ${expectedTag}`;

    const expectedErrorMessage = "You can't give more points than 500.";
    const expectedResponse = `{"errors":[{"source":{"pointer":"/data/attributes/base"},"detail":"${expectedErrorMessage}"}]}`;
    const expectedError = { response: { request: { response: expectedResponse } } };

    const mockCreateComment = jest.fn(
      () =>
        new Promise((resolve, reject) => {
          reject(expectedError);
        }),
    );
    CommentService.createComment.mockImplementation(mockCreateComment);

    const { getByTestId } = render(
      <Context.Provider value={{ dispatch }}>
        <SendCommentForm onSuccess={() => {}} bonusId={expectedBonusId} />,
      </Context.Provider>,
    );
    const textArea = getByTestId('test-textarea');

    fireEvent.change(textArea, { target: { value: expectedCommentText } });

    const form = getByTestId('test-comment-form');
    const errorContainer = getByTestId('test-error-container');

    // Act
    fireEvent.submit(form);

    // Assert
    await wait(() => {
      expect(CommentService.createComment).toHaveBeenCalledWith(expectedCommentText, expectedBonusId);
      expect(errorContainer).toHaveTextContent(expectedErrorMessage);
    });
  });
});

describe('HelperIcon', () => {
  const dispatch = jest.fn();
  const expectedBonusId = '555';

  test('Should change the className on user change in textarea value', () => {
    // Arrange
    const { getByTestId } = render(
      <Context.Provider value={{ dispatch }}>
        <SendCommentForm onSuccess={() => {}} bonusId={expectedBonusId} />,
      </Context.Provider>,
    );
    const textArea = getByTestId('test-textarea');
    const expectedCommentText = '#create-awesomeness';

    // Act
    fireEvent.change(textArea, { target: { value: expectedCommentText } });

    const commentForm = getByTestId('test-comment-form');
    // Assert
    expect(commentForm).toMatchSnapshot();
  });
});
