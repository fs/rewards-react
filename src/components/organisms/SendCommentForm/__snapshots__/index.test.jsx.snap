// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelperIcon Should change the className on user change in textarea value 1`] = `
.c4 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  overflow: hidden;
  background-color: #d8d8d8;
  margin-right: 10px;
  cursor: pointer;
  -webkit-transition: box-shadow 0.08s linear;
  transition: box-shadow 0.08s linear;
  outline: none;
  border: 0;
}

.c4:hover,
.c4.active {
  background-color: rgba(0,85,112,0.7);
}

.c4:focus {
  box-shadow: 0 1px 3px 0 rgba(60,64,67,0.302),0 4px 8px 3px rgba(60,64,67,0.149);
}

.c4:last-child {
  margin-right: 0;
}

.c2 {
  background-color: #63bc36;
  color: #fff;
  border-radius: 100px;
  height: 40px;
  line-height: 40px;
  padding: 0 20px;
  font-size: 16px;
  font-weight: bold;
  border: none;
}

.c2:hover,
.c2:focus {
  outline: none;
  background-color: #4f962b;
}

.c0 {
  width: 100%;
  background-color: #f7f7f7;
  border-radius: 5px;
  margin: 0 0 20px;
  padding: 12px 20px;
  box-shadow: 0 1px 2px 0 rgba(0,0,0,0.1);
  outline: none;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  overflow: hidden;
  height: 0;
  -webkit-transition: height 500ms;
  transition: height 500ms;
}

<form
  class="c0"
  data-testid="test-comment-form"
  tabindex="0"
>
  <div
    data-testid="test-textarea-wrapper"
  >
    <div
      class="rta  "
    >
      <textarea
        class="rta__textarea autocomplete-textarea autocomplete-textarea--comment"
        data-testid="test-textarea"
        name="bonus-text"
        placeholder="Add a comment"
      >
        #create-awesomeness
      </textarea>
    </div>
  </div>
  <div
    data-testid="test-error-container"
  />
  <div
    class="c1"
  >
    <button
      class="c2"
      data-testid="test-button"
    >
      Add
    </button>
    <div
      class="c3"
    >
      <button
        class="c4"
        type="button"
      >
        <img
          alt="Points"
          src="[object Object]"
        />
      </button>
      <button
        class="active c4"
        type="button"
      >
        <img
          alt="Hashtag"
          src="[object Object]"
        />
      </button>
    </div>
  </div>
</form>
`;
